release: define SCYLLA_BUILD_MODE_STR by stringifying SCYLLA_BUILD_MODE

bhalevy opened the pull request:
Currently SCYLLA_BULD_MODE is defined as a string by the cxxflags
generated by configure.py.  This is not very useful since one cannot use
it in a @if preprocessor directive.

Instead, use -DSCYLLA_BULD_MODE=release, for example, and define a
SCYLLA_BULD_MODE_STR as the dtirng representation of it.

In addition define the respective
SCYLLA_BUILD_MODE_{RELEASE,DEV,DEBUG,SANITIZE} macros that can be easily
used in @ifdef (or #ifndef :)) for conditional compilation.

The planned use case for it is to enable a task_manager test module only
in non-release modes.

Signed-off-by: Benny Halevy <bhalevy@scylladb.com>

avikivity commented on build_mode.hh:
```diff
@@ -0,0 +1,27 @@
+#ifndef stringify
+#define _stringify(x) #x
+#define stringify(x) _stringify(x)
+#endif
```
It's not good practice to use lowercase names for macros, it can lead to strange results if they are used in an unexpected context.

bhalevy replied:
Fixed in da321a894

bhalevy commented:
In da321a894: renamed stringification macros to `STRINGIFY_MACRO` and `STRINGIFY_VALUE`.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1915/

bhalevy commented:
ping @nyh

Deexie commented on build_mode.hh:
```diff
@@ -0,0 +1,31 @@
+#define SCYLLA_BUILD_MODE_STR STRINGIFY_MACRO(SCYLLA_BUILD_MODE)
+
+#if SCYLLA_BUILD_MODE == release
+#define SCYLLA_BUILD_MODE_RELEASE
```
Currently `SCYLLA_BUILD_MODE_RELEASE` is always defined regardless of the actual build mode.

bhalevy replied:
Fixed in v2

bhalevy commented:
In v2 (6b48a9f): Fixed `#if` evaluation

bhalevy commented:
@Deexie I converted this PR to a draft until you verify and approve the patch. Thanks

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1947/

bhalevy commented:
In v3 ([0882714](https://github.com/scylladb/scylladb/commit/0882714ee9252963dfcaf605a50c693787f8d329)):
- pass `-DSCYLLA_BUILD_MODE=...` to all files
- Added `#error` if `SCYLLA_BUILD_MODE` is not defined in build_mode.hh
- Support also coverage build mode
- Reorder modes by the same order they are defined in configure.py
- Added `#error` if SCYLLA_BUILD_MODE is unrecognized

bhalevy commented:
Since 765d2f5 was merged, the fix on top of it was submitted as #11387

Deexie reviewed (commented):
Now everything works as expected.
